### ISSUE TITLE
Behavior of the sun in bedrock

### ISSUE THREAD

**Issue by ghost** [juancarloscp52/BedrockIfy#110]
https://youtu.be/iymEkILDHAI
The sun can me the color of the sky in bedrock

**Comment by juancarloscp52** [comment]
I'm already tracking this feature as it was already posted in curseforge comments, I'll take a try as soon as I can

**Comment by lonefelidae16** [comment]
Sorry for the late reply. Your link is not available, so I hope the feature you mentioned matches what I think it does.

Depending on the sun angle and camera angle, the sky becomes darker. This is not that affects other rendering system (like auto exposure).

I’m trying to implement this feature and it goes well at this time, but I’ve found that it has some “side effect”s. I’m wondering whether to create a new PR.

**TL;DR**

Please check out this tree. ~[lonefelidae16:feat/sun-glare](https://github.com/lonefelidae16/BedrockIfy/tree/feat/sun-glare)~
Edit: successfully merged into upstream.

https://user-images.githubusercontent.com/74746227/217961309-d6f9e578-f860-493f-ad47-1a463e6281e4.mp4

This includes the main feature, the test code to be compatible with Iris, and updated project dependencies.

----

**Issues**

1. **There are no helper methods to find the sun vector <sup>(I want the vector, not angle)</sup>.**
There is a helper method to get the sky angle, but I found that I needed to calculate myself where is the sun. Also where I’m looking at.

1. **It is hard to make compatible with shader mods (e.g. [Iris Shaders](https://irisshaders.net/)).**
If any shader mods are present, the expected behavior is for this feature to be disabled. However, if the shader supports a “Vanilla” rendering <sup>(**without** external shaderpacks)</sup>, I thought that I wanted this feature to be enabled. The only exception is that this feature is included into `bedrockShading` package and therefore is disabled when optifabric exists.

----

**Solutions and Side Effects**

In order to resolve above, I include following:

1. **Calculate the sun vector and the camera vector.**
It’s not a difficult one, but is incompatible with external shaderpacks which can change the sun angle.

1. **Use Reflection to get the shader state.**
The cost of maintaining the code is too high to use Reflection. Therefore, I thought that I could resolve this by introducing a testing mechanism provided by JUnit. To implement, this project needs to be upgraded to fabric-loom 1.0 or above. Of course this is not limited to Iris; I’ve tried to make it extensible.

For more details, please check my commit log.

----

**In Creating a Pull Request**

I’m not sure if this is a feature that should be implemented using Reflection.

Had this been a project I managed, I would have gladly implemented it, but I am just a contributor. One factor is that if I were in the opposite position, I would be baffled by such “heavy” codes offering.

So, I also think it might be better to disable this feature when shader mods are present (by BedrockIfyMixinPlugin.java at commit [86cf6d6](https://github.com/lonefelidae16/BedrockIfy/commit/86cf6d6ed6cd12e923e76f9a7be91168520c50e8#diff-e3bb6688811f72fc89dd9c0f79390a3bc420108ce83ae21db0026309357523f8)). This way, there is no need to write test code and the amount of code is reduced, but cannot support switching Vanilla rendering.

Which do you think is better? Please let me know your opinion.

**Comment by juancarloscp52** [comment]
I did also try to implement this feature a the time, but was unsuccessful retrieving sun angle consistently.

This implementation is great! About the shader issue... I don't mind having reflection on the codebase (although I'll need to take a deep look at the implementation and understand it before merging a PR) in order to make it compatible with shaders. But I'm unsure if we should implement this feature also for shaders, my reasoning being that this is something that probably can be done through shaders so it might cause other incompatibilities by applying the effect twice (or a similar effect + our own one), or changing the intended look of a shader. On the other hand, I understand that some users might want this functionality to work on shaders that do not implement this kind of "auto exposure".

Also, If we were to include the implementation with reflection, how hard do you think will it be to maintain compatibility with each new Iris update?

**Comment by lonefelidae16** [comment]
Thank you for your reply!

> But I'm unsure if we should implement this feature also for shaders, my reasoning being that this is something that probably can be done through shaders so it might cause other incompatibilities by applying the effect twice (or a similar effect + our own one), or changing the intended look of a shader. On the other hand, I understand that some users might want this functionality to work on shaders that do not implement this kind of "auto exposure".

Oh, the possibility of applying this twice... I hadn’t thought that far ahead. I can understand that some users would like to have this feature enabled regardless of shaders are present. Once we start thinking about compatibility... I might get a headache. xD

> Also, If we were to include the implementation with reflection, how hard do you think will it be to maintain compatibility with each new Iris update?

Yes, that’s it. The test code is very useful for this purpose, it makes us be able to notice following:

* Has changed the class name.
* Has changed the method name.
* Has changed the access modifier of the method.
* Has changed the argument types, the number of arguments.
* Has changed the return type.

They can catch at compile time when executed the Gradle `testCompile`  or `test` task (by IrisCompatTest.java at commit [36f8377](https://github.com/lonefelidae16/BedrockIfy/commit/36f8377e0888c5b7e172baf744f82a4bb522f349#diff-1e9cac9a4d9ccf56d707f631e108bdd04aa4b4ca691b2045ab981462a9e2c7e6)). If some errors appeared, just change the string defined in [BedrockSunGlareShading#MOD_ID_CLASS_MAP](https://github.com/lonefelidae16/BedrockIfy/commit/9f6d0eaf3f2936f305c1d215658ba78cf83db256#diff-84b2e5b67c2592b3cf8502d22473feef44a680d02adeaf7b24e617e5c260986d) correctly. I think it considerably reduces the difficulty.

However, if the return value of the target method for determining shader state has changed, we must to check it manually (e.g. boolean has inverted).

**Comment by juancarloscp52** [comment]
yeah the unit testing is a great idea for ensuring compatibility. If you think we'll be able to maintain this feature through future versions, looks good to me then. In the worst case scenario this part of the feature can always be rolled back or temporally disabled. Feel free to open a PR. And of course... Thank you so much!

**Comment by lonefelidae16** [comment]
Thank you, too. It’s thanks to your awesome project I enjoy coding very much. <3
And... I feel sorry that I made bugs so often, I will be careful about it.

I would like to help maintain as possible. And of course, when a new release of Minecraft comes.

I found that I can make this feature a bit more developer friendly and user friendly, so I’ll add some commits and get ready to create a PR.
